OD-259 [Feature] Add support for date comparison filters

AndrRyaz opened the pull request:
@sofiiakvasnevska 
## Issue
OD-259 https://weboo.atlassian.net/browse/OD-259

## Description

- Merged all filter logic's correctly
- Added new optgroup for logic options
- Added new logics : 'Date is', 'Date is before' 'Date is after' 'Date is between'
- Added a filtering data between 2 dates
- Added an offset for current date & time

## Screenshots/screencasts
Filter logic: Date
http://somup.com/crVOlmogQq
Filter logic: Date is between
http://somup.com/crVOIcog6N
Sorry my recorder cannot record more than 5 min videos.
## Backward compatibility
This change is fully backward compatible.
## Reviewers
@upplabs-alex-levchenko @ivandevupp @YaroslavOvdii

tonytlwu commented on js/interface.js:
```diff
@@ -231,19 +240,47 @@ var DynamicLists = (function() {
             case 'logic':
-              var hideValueFields = ['empty', 'notempty', 'between'].indexOf(value) !== -1;
+              var hideValueFields = logicOptions.indexOf(value) !== -1;
+              var areLogicDate = ['dateis', 'datebefore', 'dateafter'].includes(value);
```
```suggestion
              var isDateLogic = ['dateis', 'datebefore', 'dateafter'].includes(value);
```

tonytlwu commented on js/interface.js:
```diff
@@ -231,19 +240,47 @@ var DynamicLists = (function() {
+              var isAddNumberFieldFrom = ['today', 'now'].includes($('#date-field-from-' + id).val());
+              var isAddNumberFieldTo = ['today', 'now'].includes($('#date-field-to-' + id).val());
+
               var isLogicComparison = value === 'between';
```
```suggestion
              var isBetween = value === 'between';
```

tonytlwu commented on js/interface.js:
```diff
@@ -231,19 +240,47 @@ var DynamicLists = (function() {
+              var isAddNumberFieldTo = ['today', 'now'].includes($('#date-field-to-' + id).val());
+
               var isLogicComparison = value === 'between';
+              var isDateBetweenLogic = value === 'datebetween';
```
```suggestion
              var isDateBetween = value === 'datebetween';
```

tonytlwu commented on js/interface.js:
```diff
@@ -231,19 +240,47 @@ var DynamicLists = (function() {
+              var hideValueFields = logicOptions.indexOf(value) !== -1;
+              var areLogicDate = ['dateis', 'datebefore', 'dateafter'].includes(value);
+
+              var isAddNumberField = ['today', 'now'].includes($('#date-field-' + id).val());
```
```suggestion
              var canAddNumber = ['today', 'now'].includes($('#date-field-' + id).val());
```

tonytlwu commented on js/interface.js:
```diff
@@ -231,19 +240,47 @@ var DynamicLists = (function() {
+              var areLogicDate = ['dateis', 'datebefore', 'dateafter'].includes(value);
+
+              var isAddNumberField = ['today', 'now'].includes($('#date-field-' + id).val());
+              var isAddNumberFieldFrom = ['today', 'now'].includes($('#date-field-from-' + id).val());
```
```suggestion
              var canAddNumberFrom = ['today', 'now'].includes($('#date-field-from-' + id).val());
```

tonytlwu commented on js/interface.js:
```diff
@@ -743,21 +776,32 @@ var DynamicLists = (function() {
 
-        $('#value-type-field-from-' + item.id).val(item.valueType.from);
-        $('#value-type-field-to-' + item.id).val(item.valueType.to);
+        if (item.logic === 'datebetween') {
```
Use `else if`s to connect these `if` clauses

tonytlwu commented on js/interface.js:
```diff
@@ -23,6 +23,15 @@ var DynamicLists = (function() {
       ]
     }
   ];
+  var logicOptions = [
```
I don't understand what `logicOptions` is and why it only has these 6 logic operators when we have a lot more. It needs to be clearer.

tonytlwu replied:
@sofiiakvasnevska This has not been resolved.

tonytlwu commented on templates/interface/filter-panels.interface.hbs:
```diff
@@ -141,5 +194,89 @@
+              </optgroup>
+            </select>
+            <span class="icon fa fa-chevron-down"></span>
+            <span class="select-value-proxy">-- Please wait</span>
```
What's this for?

```suggestion
```

tonytlwu commented on templates/interface/filter-panels.interface.hbs:
```diff
@@ -141,5 +194,89 @@
+              </optgroup>
+            </select>
+            <span class="icon fa fa-chevron-down"></span>
+            <span class="select-value-proxy">-- Please wait</span>
```
```suggestion
```

tonytlwu commented on templates/interface/filter-panels.interface.hbs:
```diff
@@ -58,13 +58,62 @@
+                </optgroup>
+              </select>
+              <span class="icon fa fa-chevron-down"></span>
+              <span class="select-value-proxy">-- Please wait</span>
```
```suggestion
```

tonytlwu commented on templates/interface/filter-panels.interface.hbs:
```diff
@@ -58,13 +58,62 @@
+              </optgroup>
             </select>
             <span class="icon fa fa-chevron-down"></span>
             <span class="select-value-proxy">-- Please wait</span>
```
```suggestion
```

tonytlwu commented on templates/interface/filter-panels.interface.hbs:
```diff
@@ -58,13 +58,62 @@
+          </div>
+          <div class="col-sm-8">
+            <label for="date-field-{{id}}" class="select-proxy-display">
+              <select name="date-field-{{id}}" id="date-field-{{id}}" data-field="date" filter-item-id="{{id}}" data-label="-- Select an option" class="hidden-select form-control">
```
What's the `data-label` for?

```suggestion
              <select name="date-field-{{id}}" id="date-field-{{id}}" data-field="date" filter-item-id="{{id}}" class="hidden-select form-control">
```

tonytlwu commented on js/interface.js:
```diff
@@ -2611,11 +2667,33 @@ var DynamicLists = (function() {
+          var valueNumberFrom = $('#number-field-from-' + item.id).val();
+          var valueNumberTo = $('#number-field-to-' + item.id).val();
+
+          item.filterModifier.to = valueDateTo;
```
When the logic is `between`, we use the filter `item.value = { from: valueFrom, to: valueTo }`.

I think we can do the same here with `databetween`:

```js
ite.value = {
  from: { value: 'today' }, // If there's no offset required, don't add the offset property
  to: { value: 'todayadddays', offset: 2 } // Parse the offset value into a number as specified
}
```

tonytlwu commented on js/utils.js:
```diff
@@ -458,6 +458,130 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+      case 'today':
+        return {
+          today: moment().format('DD/MM/YYYY'),
+          formatedDate: moment(options.date).format('DD/MM/YYYY')
```
I don't understand what the requirement is to format it into a string, and more important, why this format.

AndrRyaz replied:
@tonytlwu.

When searching for how to correctly compare dates using the moment js, I found more than one option for comparing strings, which I was surprised myself, but it works correctly.
The use of formats was necessary in order to correctly compare dates with an ordered time shift.
For example:
the record in the table is equal to 26 April 2021 16:30
our current date together with the time of April 26, 2021 12:35

to use the shift by the hour, I need to disregard the minutes, as I understood because there will be no coincidence, since the minutes differ, based on this, the Date and check check will almost always be false. But I can also be mistaken in that I have gone too deep into the intricacies of checking the time.

tonytlwu replied:
1. You can use the functions available in moment for comparison. See https://momentjs.com/docs/#/query/
2. If you want to ignore minutes, then set the minutes to 0 rather than ignoring it through string comparison

For example:

```
moment().startOf('hour') // This sets everything that's less than an hour to 0
```

AndrRyaz replied:
@tonytlwu, okay i'll finish all your comments due next review.

AndrRyaz commented:
@tonytlwu hello, I added new logic and necessary fixes, I also added a new package that helps the moment.js work with timezones, i added him into widget.json file in assets

tonytlwu commented on js/utils.js:
```diff
@@ -460,6 +611,19 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
           return splitByCommas(filter.value).includes(rowData);
```
`Array.prototype.includes()` [doesn't worked in IE](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes).

AndrRyaz commented on js/interface-lists.js:
```diff
@@ -457,8 +457,20 @@ function attahObservers() {
         if (widgetData.filterOptions.length) {
           var filterError = [];
           var filterFieldValues = [];
+          var logicOptions = [
```
@tonytlwu hello sorry for this overlooked point, I thought it was solved in a previous PR - https://github.com/Fliplet/fliplet-widget-dynamic-lists/pull/519#discussion_r564663657
We discussed the same, but in a different file
Now I have changed to the same name and the logic of this array in this file carries the same logic as in the one I indicated above

AndrRyaz commented:
@tonytlwu ,fixed

tonytlwu commented on js/interface.js:
```diff
@@ -2610,11 +2693,48 @@ var DynamicLists = (function() {
+        if (item.logic === 'datebetween') {
+          var valueDateFrom = $('#date-from-' + item.id).val();
+          var valueNumberFrom = $('#number-field-from-' + item.id).val();
+          var isTimeZoneFrom = $('#enable-timezone-from-' + item.id).is(':checked');
```
```suggestion
          var useDeviceTimezoneFrom = $('#enable-timezone-from-' + item.id).prop('checked');
```

tonytlwu commented on js/utils.js:
```diff
@@ -415,6 +415,162 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
   }
 
+  function getDateModifiedValues(options) {
+    var isEntryContainsTime = /^([0-1][0-9]|[2][0-3]):([0-5][0-9])$/.test(options.date);
```
1. Set up `timestamp` variable to simplify the code below.
2. Based on your regex, it's not just that the timestamp "contains time", it contains "only" time. So, use a variable name that's more specific to its purpose.

```suggestion
    var timestamp = options.date;
    var timeOnly = /^([0-1][0-9]|[2][0-3]):([0-5][0-9])$/.test(timestamp);
```

tonytlwu commented on js/utils.js:
```diff
@@ -415,6 +415,162 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+        return {
+          today: moment().startOf('day'),
+          inputDate: isEntryContainsTime
+            ? moment(options.date, 'HH.mm').startOf('day')
```
Why have you chosen to use `HH.mm` with `.` and not the standard `:` syntax?

```suggestion
            ? moment(options.date, 'HH:mm').startOf('day')
```

Also, based on the requirements highlight below, if the entry only contains time, it's not meant to be valid, therefore fails the filter check. Why set the input date as `moment().startOf('day')`?

![image](https://user-images.githubusercontent.com/290733/108637212-af903980-7481-11eb-8e01-3928676004a5.png)

AndrRyaz replied:
Fixed

tonytlwu commented on js/utils.js:
```diff
@@ -415,6 +415,162 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+      case 'now':
+        return {
+          today: getDate().startOf('minute'),
+          inputDate: isEntryContainsTime
```
I don't see any checks to verify whether a value that's used for time comparison contains actual time.

![image](https://user-images.githubusercontent.com/290733/108637287-faaa4c80-7481-11eb-944c-8cf53ecb0b5d.png)

AndrRyaz replied:
Fixed

tonytlwu commented on js/utils.js:
```diff
@@ -415,6 +415,162 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+
+      switch (options.condition) {
+        case 'dateis':
+          return moment(result.inputDate).isSame(result.today);
```
It looks like `result.today` is calculated statically, i.e. does not depend on the value of each record, but is being creating from scratch every single time in the `getDateModifiedValues()` function, e.g. in `case 'now'`, this value is `today: getDate().startOf('minute')`. Why does this need to be created every single time instead of using a cached value?

AndrRyaz replied:
I cannot move this variable to another place because in this case we can have a scenario in between comparison where we will have in From value with using time zone and in To value without using time zone

tonytlwu replied:
Assuming there is a filter that uses date comparison, this is my breakdown:

1. `isDateMatches()` is getting called as many times as the the number of records
2. `getDateModifiedValues()` is getting called as many times as there are the number of records
3. The `getDateModifiedValues()` runs something like `getDate().add('minute', smartParseFloat(options.modifierValue)).startOf('minute')` as many times as the number of records. However, `getDate().add('minute', smartParseFloat(options.modifierValue)).startOf('minute')` should produce the same output every single time. If the data source has 10,000 entries, we don't need to compute this value 10,000 times.

You need to fix this problem.

I suggest making sure that the value you pass to `.isSame()`, `.isBefore()` etc. only gets calculated only once for each filter.

tonytlwu commented on js/utils.js:
```diff
@@ -1311,7 +1475,10 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+        filterModifier: option.filterModifier,
+        valueType: option.valueType,
+        value: option.value,
+        timezone: option.timezone
```
Where does the `option.timezone` value come from?

AndrRyaz replied:
In this place this variable is unnecessary

tonytlwu commented on js/utils.js:
```diff
@@ -1311,7 +1475,10 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
         column: option.column,
         condition: option.logic,
-        value: option.value
+        filterModifier: option.filterModifier,
```
I don't understand the purpose of `filterModifier`, how it's used and why it's necessary.

AndrRyaz replied:
This variable has many necessary values such as offset, oseTimezone and others. They are required for every filter.

AndrRyaz commented on js/interface.js:
```diff
@@ -2610,11 +2693,48 @@ var DynamicLists = (function() {
+
         if (item.logic === 'empty' || item.logic === 'notempty') {
-          item.valueType = null;
+          item.valueType = '';
```
I changed the value to an empty string due to the fact that all values of this type with a false value have an empty string, for the sake of purity in my opinion, this line was distinguished by the fact that it assigned a different value, although it carries the same functionality as all the others. I have tested and this change has no consequences. But I can change the given variable in reverse

In the case of null, it means that we do not know what a value is, it may or may not exist, we simply do not know. 
An empty string means that we know what the value is and that it is nothing.

AndrRyaz replied:
I changed this value for consistency. I tested everything works correctly

AndrRyaz commented on js/interface.js:
```diff
@@ -221,6 +230,7 @@ var DynamicLists = (function() {
           item.value = '';
           item.valueField = 'Value';
           item.valueType = 'enter-value';
+          item.isDeviceTimeZoneEnabled = false;
```
Fixed

AndrRyaz commented on js/interface.js:
```diff
@@ -2610,11 +2693,48 @@ var DynamicLists = (function() {
+          default: {
+            value: $('#date-field-' + item.id).val(),
+            offset: $('#number-field-' + item.id).val() ? $('#number-field-' + item.id).val() : null,
+            isDeviceTimeZoneEnabled: $('#enable-timezone-default-' + item.id).is(':checked')
```
fixed

AndrRyaz commented on js/interface.js:
```diff
@@ -232,19 +242,59 @@ var DynamicLists = (function() {
           var type = $(this).data('field');
           var $selector = $(this).parents('.filter-panel');
           var id = $(this).attr('filter-item-id');
+          var isAddNumberFieldDefault = ['today', 'now'].indexOf(value) !== -1;
```
Fixed

AndrRyaz commented on js/interface.js:
```diff
@@ -2610,11 +2693,48 @@ var DynamicLists = (function() {
+            default: {
+              value: $('#date-field-' + item.id).val(),
+              offset: $('#number-field-' + item.id).val() ?  $('#number-field-' + item.id).val() : null,
+              isDeviceTimeZoneEnabled: $('#enable-timezone-default-' + item.id).is(':checked')
```
Fixed

AndrRyaz commented on js/interface.js:
```diff
@@ -2610,11 +2693,48 @@ var DynamicLists = (function() {
+          item.filterModifier = {
+            default: {
+              value: $('#date-field-' + item.id).val(),
+              offset: $('#number-field-' + item.id).val() ?  $('#number-field-' + item.id).val() : null,
```
Fixed

AndrRyaz commented:
@tonytlwu Fixed

tonytlwu commented on js/interface.js:
```diff
@@ -2610,11 +2696,47 @@ var DynamicLists = (function() {
+          var useDeviceTimezoneTo = $('#enable-timezone-to-' + item.id).prop('checked');
+
+          item.filterModifier = {
+            default: {
```
I assume the `datebewteen` logic requires modifiers for `from` and `to`. Why does it need values for the `default` modifier?

If it's not needed, we shouldn't be saving unnecessary data, which could increase the widget data and page size.

tonytlwu commented on js/interface.js:
```diff
@@ -2610,11 +2696,47 @@ var DynamicLists = (function() {
+          var valueNumberTo = $('#number-field-to-' + item.id).val();
+          var useDeviceTimezoneTo = $('#enable-timezone-to-' + item.id).prop('checked');
+
+          item.filterModifier = {
```
```suggestion
          item.dateFilterModifiers = {
```

tonytlwu commented on js/interface.js:
```diff
@@ -2610,11 +2696,47 @@ var DynamicLists = (function() {
           item.value = item.fieldValue;
         }
 
+        item.filterModifier = {
```
Why is `filterModifier` set for all filter options when the modifier isn't necessarily always needed?

tonytlwu commented on js/utils.js:
```diff
@@ -1311,6 +1478,8 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
       return {
         column: option.column,
         condition: option.logic,
+        filterModifier: option.filterModifier,
```
```suggestion
        dateFilterModifiers: option.dateFilterModifiers,
```

tonytlwu reviewed (changes requested):
Does not meet coding standard.

AndrRyaz commented:
@tonytlwu hi, added changes

tonytlwu commented on js/interface.js:
```diff
@@ -238,13 +240,47 @@ var DynamicLists = (function() {
               break;
 
+            case 'date':
+              var dateField = $('#date-field-' + id).val();
```
https://github.com/airbnb/javascript/blob/master/README.md#jquery--dollar-prefix

tonytlwu commented on js/interface.js:
```diff
@@ -238,13 +248,50 @@ var DynamicLists = (function() {
+              var hideValueFields = logicOptionsWithoutValues.indexOf(value) !== -1;
+              var isDateLogic  = ['dateis', 'datebefore', 'dateafter'].indexOf(value) !== -1;
+
+              var canAddNumber = ['today', 'now'].indexOf($('#date-field-' + id).val()) !== -1;
```
@AndrRyaz Not all occurences are resolved.

tonytlwu commented on js/utils.js:
```diff
@@ -415,6 +416,191 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
+    switch (options.dateFilterModifiers) {
+      case 'today':
+        return {
+          today: moment().startOf('day'),
```
1. `moment().startOf('day')` should also go through the same caching system that the other date modifiers are using. See https://jsbench.me/q5klic37ss for how poor `moment().startOf('day')` performs.
2. `today` isn't really "today". Instead of the values `today` and `inputDate`, let's use `comparisonDate` and `entryDate` instead. `comparisonDate` is the control value for comparison and `entryDate` represents the value from each entry.
3. `getCurrentDate()` should be `getCachedDate()`. `getCachedDate()` will load the value from cache if it exists, or create the value from scratch and save it into cache before returning the value.

AndrRyaz commented:
@tonytlwu hello i added fixes to variable names and when i checked the comparison again i found 1 miss point and added fixes for it

tonytlwu reviewed (commented):
@sofiiakvasnevska Please send this to QA to test on development.

tonytlwu reviewed (approved):
@sofiiakvasnevska This is ready for testing

Risk: Medium
Affected: All LFD users
Verify: Filter settings are working correctly in Studio and apps

Squash and merge: Use PR title
